### PR TITLE
Iox #911 design document dynamic size types

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CHANGELOG.md

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Design document task in #911

### CODE REVIEW

File: doc/design/drafts/dynamic_size_types.md [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+chunk where all the contents are stored inside one chunk. Hereby is the
+Memory Manager the class which handles the distribution of shared memory chunks.
+To support different sizes the Memory Manager uses bucket allocators with
+preconfigured bucket sizes called Memory Pool.
+
+```
++--------------------- Memory Manager -------------------+
+|  MemPool with chunk size = 1000                        |
+|  +----------------+----------------+----------------+  |
+|  |  data          |                |                |  |
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:90]: I think this should be ``value``.

File: doc/design/drafts/dynamic_size_types.md [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+its local process.
+
+## Design Allocator
+
+The allocator should be designed in a manner so that it can be used inside the
+shared memory in an IPC context. This means the design has to consider that
+vtables, virtual and inheritance are not allowed as well as function pointers
+or the use of `cxx::function` or `cxx::function_ref`.
+
+Therefore this allocator concept is less flexibel and more complex as it would
```
Comment (r7vme) [inline doc/design/drafts/dynamic_size_types.md:144]:
```suggestion
Therefore this allocator concept is less flexible and more complex as it would
```

File: doc/design/drafts/dynamic_size_types.md [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+```
+
+### Usage
+
+```cpp
+cxx::vector<int, 20> a;
+a.emplace_back(123); // can be used without a preceding reserve call
+
+cxx::vector<int> c(heapAllocator);
+c.reserve(42); // reserve required
```
Comment (r7vme) [inline doc/design/drafts/dynamic_size_types.md:377]: what should happen if no reserve was called?

Comment (elfenpiff) [reply doc/design/drafts/dynamic_size_types.md:377]: In the case of the static vector it is not required and nothing will happen. For all other vectors `cxx::Expects` is called with the expectation that reserve was called before - if not `cxx::Expects` will call `std::terminate()` but this can be configured to something else later when we have a more fine grained error handling concept.

File: doc/design/drafts/dynamic_size_types.md [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+of a vector. It makes the following operations possible
+```cpp
+void f(const vector<int, 20> &a) {}
+
+vector<int> v1;
+vector<int, 20> v2;
+
+// this works thanks to implicit conversions from vector<int>
+f(v1);
+v2 = v1;
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:336]: What happens when the size of `v1` is greater than 20?

File: doc/design/drafts/dynamic_size_types.md [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
++--------------------------------------------------------+
+```
+
+### Layout With Dynamic Sized Types
+
+With dynamic sized types we move the data part outside of the container and
+add an allocator to the container which allows the container to acquire
+memory for their data part. When we assume that `MyData::image` and
+`MyData::text` require 100 bytes for their data part the memory layout could
+look like this.
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:112]: What happens when there is not enough memory for the data Part?

File: doc/design/drafts/dynamic_size_types.md (hunk 3 above)
Comment (FerdinandSpitzschnueffler) [reply doc/design/drafts/dynamic_size_types.md:377]: Can reserve fail?

File: doc/design/drafts/dynamic_size_types.md (hunk 5 above)
Comment (elfenpiff) [reply doc/design/drafts/dynamic_size_types.md:112]: I adjusted the allocator interface so that it provides the memory packed inside an expected with a custom allocator error type. This is error is forwarded via `reserve` when it fails.

File: doc/design/drafts/dynamic_size_types.md (hunk 4 above)
Comment (elfenpiff) [reply doc/design/drafts/dynamic_size_types.md:342]: It fails, a detailed explanation was added to the document.

File: doc/design/drafts/dynamic_size_types.md (hunk 3 above)
Comment (elfenpiff) [reply doc/design/drafts/dynamic_size_types.md]: The allocator returns an error via expected which is forwarded to the user when reserve fails.

**Comment by elfenpiff** [comment]
@elBoberido @FerdinandSpitzschnueffler since it is a design document only I would be happy if you can accept the PR so that it can be merged. If too many questions are opened I would be happy to add the to the design document or to move the design document into the draft folder.

File: doc/design/drafts/dynamic_size_types.md [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  +---+---+---+---+---+---+---+---+---+---+---+---+---+ |
++--------------------------------------------------------+
+```
+
+### Layout With Dynamic Sized Types
+
+With dynamic sized types we move the data part outside of the container and
+add an allocator to the container which allows the container to acquire
+memory for their data part. When we assume that `MyData::image` and
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:110]: Please describe this in more detail. The data part will now be stored in separate chunks and the management part will be inside one chunk, now including (a pointer to) the allocator, right?

File: doc/design/drafts/dynamic_size_types.md [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+        // suggested allocator
+
+  data; // contains the user data which is stored in the container
+}
+```
+
+### Static Layout
+
+When we require a sample with our publisher and write `MyData` we acquire a
+chunk where all the contents are stored inside one chunk. Hereby is the
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:81]:
```suggestion
When we require a sample with our publisher and write `MyData`, we acquire a
chunk where all the contents are stored. Hereby is the
```

File: doc/design/drafts/dynamic_size_types.md [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+```
+
+This does not require any adaptions on the subscriber side since the subscriber
+already mapped this memory region which is handled by the memory manager into 
+its local process.
+
+## Design Allocator
+
+The allocator should be designed in a manner so that it can be used inside the
+shared memory in an IPC context. This means the design has to consider that
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:140]:
```suggestion
The allocator should be designed to be used inside the
shared memory within an IPC context. This means the design has to consider that
```

File: doc/design/drafts/dynamic_size_types.md [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+vtables, virtual and inheritance are not allowed as well as function pointers
+or the use of `cxx::function` or `cxx::function_ref`.
+
+Therefore this allocator concept is less flexible and more complex as it would
+be when everything would run in the same process since we have to use relative
+pointer and have to allocate everything on the stack.
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:146]:
```suggestion
pointers and have to allocate everything on the stack.
```

File: doc/design/drafts/dynamic_size_types.md [hunk 10]
```diff
@@ -0,0 +1,562 @@
...
+               |   - expected<relative_ptr, AllocatorError> allocate(uint64_t) |
+               |   - void free(relative_ptr chunk)                             |
+               |   - template<AnotherAllocator>                                |
+               |     bool isMoveCompatible(const AnotherAllocator & rhs)       |
+               |                                                               |
+               |   # relative_ptr m_ptrToAllocator                             |
+               +---------------------------------------------------------------+
+```
+The `Allocator` described in this diagram is similar to a C++20 concept which can
+be verified at compiletime without inheritance. The verification can be realized
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:207]:
```suggestion
be verified at compile time without inheritance. The verification can be realized
```

File: doc/design/drafts/dynamic_size_types.md [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+   Stack based containers do not require a `reserve()` call beforehand but
+   should provide this call as well for compatibility reasons.
+
+   * It is not possible to provide solely the capacity (`reserve` parameter) via
+     the constructor of the container since frameworks like ROS2 cannot forward
+     such arguments when for instance `loan` was called. But it is possible to
+     provide it as an alternative.
+
+6. The allocator type should not change the underlying container type, e.g. the
+   allocator should not provided as template argument, otherwise
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:300]:
```suggestion
   allocator should not be provided as template argument, otherwise
```

File: doc/design/drafts/dynamic_size_types.md [hunk 12]
```diff
@@ -0,0 +1,562 @@
...
+   ```
+   After a copy- or move assignment operation the allocator which was provided
+   in the constructor will not change.
+
+   When the copy- or move-constructor is used the allocator of the origin shall
+   be used in the newly created container.
+
+   * Real copy- and move operations could have some unwanted side effects. Lets
+     assume one would like to copy/move a shared memory located vector into a local
+     cache. If the allocator would change the container has to allocate another
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:327]:
```suggestion
     cache. If the allocator would change, the container has to allocate another
```

File: doc/design/drafts/dynamic_size_types.md [hunk 13]
```diff
@@ -0,0 +1,562 @@
...
+       |  - vector& operator=(vector<T> &&)          |
+       |                                             |
+       |                                             |
+       |  // same API as current cxx::vector         |
+       |                                             |
+       |  # uint8_t m_data[Capacity]                 |
+       |  # RangeAllocator m_allocator               |
+       +---------------------------------------------+
+```
+All non stack based version of cxx containers will not have constructors similar
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:380]:
```suggestion
All non stack based versions of cxx containers will not have constructors similar
```

File: doc/design/drafts/dynamic_size_types.md [hunk 14]
```diff
@@ -0,0 +1,562 @@
...
+// this works since vector<int, 10> is also a vector<int> and can be converted
+v2 = v1;
+v1 = v2; // this can fail when v1 capacity is smaller then v2.size, previously 
+         // it was only possible when v1.capacity < v2.capacity
+f(v1);
+f(v2);
+```
+
+The only restriction are functions which require a stack based vector of a 
+specific size as non const reference. This is usally the case when some function
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:420]:
```suggestion
The only limitation is functions which require a stack based vector of a 
specific size as non const reference. This is usally the case when some function
```

File: doc/design/drafts/dynamic_size_types.md [hunk 15]
```diff
@@ -0,0 +1,562 @@
...
+allocator. The internal structure of those containers looks similar to the
+structure of their `cxx` pendants
+```cpp
+template<typename T>
+class vector {
+  //...
+  T * memoryStorageOfVectorElements;
+};
+```
+The pointer hereby is an absolut pointer pointing to a piece of memory in the
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:497]:
```suggestion
The pointer hereby is an absolute pointer pointing to a piece of memory in the
```

File: doc/design/drafts/dynamic_size_types.md [hunk 16]
```diff
@@ -0,0 +1,562 @@
...
+     struct MyData {
+       cxx::vector<int> value;
+     };
+
+     dataOnSharedMemory = subscriber->take();
+     MyData cache{.value = cxx:vector<int>(HeapAllocator)};
+     cache = dataOnSharedMemory; // would suddenly allocate shared memory
+     ```
+
+The draft we provide here is using the `cxx::vector` but the techniques
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:343]:
```suggestion
The provided draft uses the `cxx::vector` but the techniques
```

File: doc/design/drafts/dynamic_size_types.md [hunk 17]
```diff
@@ -0,0 +1,562 @@
...
+       |  // same API as current cxx::vector         |
+       |                                             |
+       |  # uint8_t m_data[Capacity]                 |
+       |  # RangeAllocator m_allocator               |
+       +---------------------------------------------+
+```
+All non stack based version of cxx containers will not have constructors similar
+to the STL. The only provided constructor has one argument which specifies the
+type of allocator one would like to use for that container. The default constructor
+shall construct a vector which has a capacity of 0.
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:383]: Can you please elaborate? Why can't we have typical c'tors with the allocator as last parameter with a default value, e.g. the heap allocator? This would be more user friendly.

File: doc/design/drafts/dynamic_size_types.md [hunk 18]
```diff
@@ -0,0 +1,562 @@
...
+
+1. Send via untyped publisher and provide the size at runtime to the publisher.
+   But defining message types with multiple dynamic sized containers would be
+   a real challenge when they should be constructed inside of this sample.
+
+2. One can perform a worst case estimation at compile time and declare
+   a type with container members which fullfil those estimations.
+   This could lead to a lot of wasted memory and if the worst case estimations
+   have to be adjusted all the applications which are using this type may have
+   to be adjusted as well since the capacity is part of the type.
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:17]: These estimations are also often hard to come by for more complicated structures, e.g. specific trees with internal meta data nodes etc. The estimates then are very pessimistic and as stated here lead to a lot of wasted memory which cannot be avoided by this approach. Hence it is not really an option if memory is a concern.

File: doc/design/drafts/dynamic_size_types.md [hunk 19]
```diff
@@ -0,0 +1,496 @@
...
+       |  - vector(vector<T> &&)                     |
+       |  - vector& operator=(const vector<T> &)     |
+       |  - vector& operator=(vector<T> &&)          |
+       |                                             |
+       |                                             |
+       |  // same API as current cxx::vector         |
+       |                                             |
+       |  # uint8_t m_data[Capacity]                 |
+       |  # RangeAllocator m_allocator               |
+       +---------------------------------------------+
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply doc/design/drafts/dynamic_size_types.md:378]: I added this idea to the design document under open questions to keep this idea documented.

File: doc/design/drafts/dynamic_size_types.md [hunk 20]
```diff
@@ -0,0 +1,562 @@
...
+   a real challenge when they should be constructed inside of this sample.
+
+2. One can perform a worst case estimation at compile time and declare
+   a type with container members which fullfil those estimations.
+   This could lead to a lot of wasted memory and if the worst case estimations
+   have to be adjusted all the applications which are using this type may have
+   to be adjusted as well since the capacity is part of the type.
+
+3. Another approach is to use the heap and copy/serialize the data into a sample
+   before sending it with iceoryx. This would imply multiple copies since the
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:20]: Leads to one serialization and one deserialization and is optimal *if data has to be serialized anyway*.

No copies needed if done correctly, but tota effort is effort of Omega(size of data to be sent), i.e. at least linear as a copy with more complex logic (leading to increased costs compared to copies).

File: doc/design/drafts/dynamic_size_types.md (hunk 6 above)
Comment (elfenpiff) [reply doc/design/drafts/dynamic_size_types.md:112]: @FerdinandSpitzschnueffler I added a more detailed description. Can you please check it out.

File: doc/design/drafts/dynamic_size_types.md [hunk 21]
```diff
@@ -0,0 +1,685 @@
...
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  +---+---+---+---+---+---+---+---+---+---+---+---+---+ |
++--------------------------------------------------------+
+```
+
+### Layout With Dynamic Sized Types
+
+With dynamic sized types we split the container data into two parts, the
+the actual data and a management part which contains a pointer to the data part.
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:109]:
```suggestion
actual data and a management part which contains a pointer to the data part.
```

File: doc/design/drafts/dynamic_size_types.md [hunk 22]
```diff
@@ -0,0 +1,685 @@
...
+The data part lives then outside of the container and an allocator inside the
+management part of the container allows the container to acquire
+memory for their data part. When we assume that `MyData::image` and
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/drafts/dynamic_size_types.md:112]:
```suggestion
memory for its data part. When we assume that `MyData::image` and
```

File: doc/design/drafts/dynamic_size_types.md [hunk 23]
```diff
@@ -0,0 +1,562 @@
...
+   before sending it with iceoryx. This would imply multiple copies since the
+   publisher has to copy the data from the container into the sample and every
+   subscriber may has to deserialize/copy the data out of the iceoryx sample.
+
+Those obstacles can be overcome when iceoryx containers like `cxx::string`,
+`cxx::vector` or `cxx::{forward}list` support custom allocators which allocate
+memory in a part of the shared memory which is also in the process space of the
+subscriber available.
+
+To accomplish dynamic size types we require:
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:29]: To support ...

File: doc/design/drafts/dynamic_size_types.md [hunk 24]
```diff
@@ -0,0 +1,562 @@
...
+
+1. an allocator concept which supports states for allocators
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:31]: What is 1. supposed to mean?

File: doc/design/drafts/dynamic_size_types.md [hunk 25]
```diff
@@ -0,0 +1,562 @@
...
+to shared memory allocated types which would result in some implementation
+overhead. One example could be that someone analyses an image with a function
+like
+`void analyseImage(cxx::vector<int, SharedMemoryAllocator> & a)`. This function
+can now be only used with a vector which is using the `SharedMemoryAllocator`.
+One could define the allocator type as a template but this would increase the
+compile time and can make it nearly impossible to write proprietary code since
+templates require the implementation in the header. Furthermore, the function
+developer has to be aware of the implementation detail that there is something
+like a `SharedMemoryAllocator`.
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:51]: Basically the approach of https://en.cppreference.com/w/cpp/memory/polymorphic_allocator but we will have additional complications due to shared memory (across processes).

File: doc/design/drafts/dynamic_size_types.md [hunk 26]
```diff
@@ -0,0 +1,562 @@
...
+  mgmt; // contains internal variables to handle the logic and later the
+        // suggested allocator
+
+  data; // contains the user data which is stored in the container
+}
+```
+
+### Static Layout
+
+When we require a sample with our publisher and write `MyData`, we acquire a
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:80]: request or acquire a sample

File: doc/design/drafts/dynamic_size_types.md [hunk 27]
```diff
@@ -0,0 +1,562 @@
...
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  +---+---+---+---+---+---+---+---+---+---+---+---+---+ |
++--------------------------------------------------------+
+```
+
+### Layout With Dynamic Sized Types
+
+With dynamic sized types we move the data part outside of the container and
+add an allocator to the container which allows the container to acquire
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md]: important: I think we actually only should add a reference/pointer to the allocator in general.

File: doc/design/drafts/dynamic_size_types.md [hunk 28]
```diff
@@ -0,0 +1,562 @@
...
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  +---+---+---+---+---+---+---+---+---+---+---+---+---+ |
++--------------------------------------------------------+
+```
+
+This does not require any adaptions on the subscriber side since the subscriber
+already mapped this memory region which is handled by the memory manager into 
+its local process.
+
+## Design Allocator
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:139]: transpose

File: doc/design/drafts/dynamic_size_types.md [hunk 29]
```diff
@@ -0,0 +1,562 @@
...
+  |  AllocatorError [enum class] |
+  |                              |
+  |     # OUT_OF_MEMORY          |
+  |     # MEMORY_SIZE_TOO_LARGE  |
+  +------------------------------+
+
+  +---------------------------------------------------------------+
+  |  Allocator [Concept]                                          |
+  |                                                               |
+  |   - expected<relative_ptr, AllocatorError> allocate(uint64_t) |
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:161]: We should not return relaive_ptr or similar. There is no need, we can return raw pointers. Only if we want to store them for IPC use we need to consider fancy pointers. This is not the responsibility of allocate.

File: doc/design/drafts/dynamic_size_types.md [hunk 30]
```diff
@@ -0,0 +1,562 @@
...
+  |   - void free(relative_ptr chunk)                             |
+  |   - template<AnotherAllocator>                                |
+  |     bool isMoveCompatible(const AnotherAllocator& rhs)        |
+  +---+--------+--------------------------------------------------+
+      |        |
+      |  +-----+---------------------------------------------------------+
+      |  | RangeAllocator                                                |
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:168]: Probably to early for specific allocators but sure, we want to support different specializations.

If we mention the different allocators here we should explain their purpose (HeapAllocator is also not an ideal name I think).

File: doc/design/drafts/dynamic_size_types.md [hunk 31]
```diff
@@ -0,0 +1,562 @@
...
+be verified at compile time without inheritance. The verification can be realized
+with a `IsAllocatorTypeTrait` in C++14.
+The shared memory and IPC restrictions are forcing us to implement a
+VariantAllocator which provides us type independent access to the underlying
+allocator like inheritance would provide. Hereby the implementation for
+`allocate`, `free` and `isMoveCompatible` is always the same and can be
+implemented similar to the `cxx::variant` move and copy operations. This would
+make the `VariantAllocator` easier extendable since the user just has to add
+any additional allocator as type in the variadic template list. An alternative
+way of implementing this `VariantAllocator` would be with a switch statement
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:218]: switch is a bad alternative and essentially variant abstracts something like this.

File: doc/design/drafts/dynamic_size_types.md [hunk 32]
```diff
@@ -0,0 +1,562 @@
...
+   myFunkyFunction(b);
+   myFunkyFunction(c);
+   ```
+
+4. All containers must be shared memory compatible, this means no vtables and
+   virtual. Furthermore the usage of function pointers, `cxx::function` and
+   `cxx::function_ref` is forbidden in all containers and their underlying 
+   constructs.
+
+5. Before using any cxx container the user as to call `reserve()` once otherwise
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:289]: has to

File: doc/design/drafts/dynamic_size_types.md [hunk 33]
```diff
@@ -0,0 +1,562 @@
...
+   the container will have a default capacity of zero. It is not allowed to
+   call `reserve()` multiple times since this could violate the zero copy
+   guarantee.
+   Stack based containers do not require a `reserve()` call beforehand but
+   should provide this call as well for compatibility reasons.
+
+   * It is not possible to provide solely the capacity (`reserve` parameter) via
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:296]: I think for this it is also important what kind of template types we support for say `vector`.

File: doc/design/drafts/dynamic_size_types.md [hunk 34]
```diff
@@ -0,0 +1,562 @@
...
+7. The allocator should not be part of copy and move-assignment operations.
+   The container should behave as closely as possible to an STL container with
+   custom allocator.
+   ```cpp
+   cxx::vector a(HeapAllocator);
+   cxx::vector b(SharedMemoryAllocator);
+   cxx::vector c(AnotherAllocator);
+   cxx::vector d(HeapAllocator);
+
+   a = b; // the content is copied from shared memory into the heap
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:316]: I think the behavior needs to be clarified more here. I think `a` will/should deallocate its current memory (or extend it), allocate more memory if needed and then copy the data from b to this memory.

However, in general this operation needs to know something about the source memory here (and it will, since it has access to the allocator of the source `b`).

TLDR: the actual semantics need more detail here I think.

File: doc/design/drafts/dynamic_size_types.md [hunk 35]
```diff
@@ -0,0 +1,562 @@
...
+   ```
+   After a copy- or move assignment operation the allocator which was provided
+   in the constructor will not change.
+
+   When the copy- or move-constructor is used the allocator of the origin shall
+   be used in the newly created container.
+
+   * Real copy- and move operations could have some unwanted side effects. Lets
+     assume one would like to copy/move a shared memory located vector into a local
+     cache. If the allocator would change, the container has to allocate another
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:329]: Can/should the allocator of an individual vector ever change once constructed? What would be the semantics?

File: doc/design/drafts/dynamic_size_types.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply doc/design/drafts/dynamic_size_types.md:380]: Not sure what `payload_vector` is here, I probably missed something.

But we may also request the allocator from the publisher in a way. Then we could provide this allocator to the vector as sketched in the document, but of course we would have to be careful with the lifetime.

Of course the vector would now have a pointer to the allocator, which I think the proposal here tries to avoid.
I think 2. is very error-prone and cumbersome to use.

File: doc/design/drafts/dynamic_size_types.md [hunk 36]
```diff
@@ -0,0 +1,562 @@
...
+   /* error handling */
+   return;
+}
+sample->sharedMemoryVector.emplace_back(someValue);
+```
+
+
+Copy and move operations should be possible without the developer knowing
+what kind of underlying allocator is used. Additionally, it should be as efficient
+as possible. That means the move operation should be a real move operation
```
Comment (MatthiasKillat) [inline doc/design/drafts/dynamic_size_types.md:458]: We could also define that move only works if source and destination allocator are the same (and maybe other conditions hold), otherwise we could always fall back to a copy.

Another approach is simply that the destiantion takes over the memory and allocator of the source and discards its own. If the allocator is not owned by the vector (but just a fancy pointer to it), this should be well defined and cheap, at least for vector.

The source of the move will not need allocator nor data (it should not be accessed anymore, but we can make it move aware for added safety at the cost of some runtime checks whether the allocator is valid/set).

**Comment by elBoberido** [comment]
@elfenpiff can we close this. It also starts to smell ;)

If you just want to have a quick link to this PR, you can open a page on `Discussion` and link it from there.

**Comment by elBoberido** [comment]
@elfenpiff  I think there is still too much discussion ongoing and `doc/design/draft` is more for stuff which is settled and anyone could start with implementation.

Like mentioned. My proposal would be to open a page in `Discussion` and link this PR from there. Then it is still easy to find this PR.
